### ISSUE TITLE
Odd behaviour when using a virtual foreach as the sole child of a template

### ISSUE THREAD

**Issue by rampion** [knockout/knockout#1947]
Say I want to use templates to insert `td`s:

    <table data-bind="foreach:$data">
      <tr>
        <td data-bind="text:template_name"></td>
        
      </tr>
    </table>

What works is having a template insert just one `td`:

    <script type="text/html" id="a-single-cell">
      <td data-bind="text:$data"></td>
    </script>

Or having a template insert multiple `td`s, if it uses a nested template:

    <script type="text/html" id="this-works">
      
        
      
    </script>

But, what doesn't work is if the template tries to define all the `td`s itself:

    <script type="text/html" id="this-breaks">
      
        <td data-bind="text:$data">curious</td>
      
    </script>

The `td` elements are never inserted - though oddly enough their original text is.

Unless, that is, the template is hardcoded to also insert another `td` first.  That works:

    <script type="text/html" id="this-also-works">
      <td>and curiouser</td>
      
        <td data-bind="text:$data">no?</td>
      
    </script>

For completeness, the javascript:

    var cells = [ "A", "B", "C" ];
    ko.applyBindings([
      { template_name: "a-single-cell", template_data: "I'm the first row"},
      { template_name: "this-works", template_data: cells },
      { template_name: "this-breaks", template_data: cells },
      { template_name: "this-also-works", template_data: cells },
    ])

[Here's a jsfiddle of the complete bug using Knockout-3.4.0](http://jsfiddle.net/twbjv8e1/2/).

**Comment by brianmhunt** [comment]
Thanks @rampion – Is it this problem: https://github.com/knockout/knockout/issues/1759#issuecomment-91534874 ?

**Comment by rampion** [comment]
Nope, the behaviour's the same when I [manually include the `<tbody>` tag](http://jsfiddle.net/twbjv8e1/3/).

And I forgot to mention, this is Firefox-42.0 (though I've seen it in earlier ones).

**Comment by brianmhunt** [comment]
Got it.

That is weird.  And broken. :o)  Thanks for the detailed report.  Will have a look, hopefully shortly!

**Comment by rampion** [comment]
And I'm also seeing the same behaviour in Chrome-47.0.2526.73.

Complete html for testing:

    <!doctype html>
    <html lang="en">
      <head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <title>Knockout bug</title>
        <style>
          table { border-collapse: collapse }
          td { border: 1px solid black }
        </style>
      </head>
      <body>
        
        <script type="text/html" id="a-single-cell">
          <td data-bind="text:$data"></td>
        </script>
        
        <script type="text/html" id="this-works">
          
            
          
        </script>
        
        <script type="text/html" id="this-breaks">
          
            <td data-bind="text:$data">curious</td>
          
        </script>
        
        <script type="text/html" id="this-also-works">
          <td>and curiouser</td>
          
            <td data-bind="text:$data">no?</td>
          
        </script>
        <table><tbody data-bind="foreach:$data">
          <tr>
            <td data-bind="text:template_name"></td>
            
          </tr>
        </tbody></table>
        <script type="text/javascript" src="http://knockoutjs.com/downloads/knockout-3.4.0.js"></script>
        <script type="text/javascript">
          var cells = [ "A", "B", "C" ];
          ko.applyBindings([
            { template_name: "a-single-cell", template_data: "I'm the first row"},
            { template_name: "this-works", template_data: cells },
            { template_name: "this-breaks", template_data: cells },
            { template_name: "this-also-works", template_data: cells },
          ])
        </script>
      </body>
    </html>

**Comment by mbest** [comment]
`simpleHtmlParse` doesn't strip initial comments and so doesn't see the `<td>`. Another workaround would be to include jQuery so that Knockout uses its parsing code. Also if you only need to target modern browsers, you can use `<template>` instead of `<script>` for the templates.
